sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1251,16 +1271,48 @@ RTDeviceBinaryImage *getBinImageFromMultiMap(
   if (ItBegin == ItEnd)
     return nullptr;
 
-  std::vector<sycl_device_binary> RawImgs(std::distance(ItBegin, ItEnd));
-  auto It = ItBegin;
-  for (unsigned I = 0; It != ItEnd; ++It, ++I)
-    RawImgs[I] = reinterpret_cast<sycl_device_binary>(
-        const_cast<sycl_device_binary>(&It->second->getRawData()));
+  // Here, we aim to filter out all the device images from
+  // [ItBegin, ItEnd) range that are not AOT compiled for the
+  // Device (checked using info::device::architecture).
+  std::string_view ArchName = getArchName(Device);
+  std::vector<RTDeviceBinaryImage *> DeviceFilteredImgs;
+  DeviceFilteredImgs.reserve(std::distance(ItBegin, ItEnd));
+  for (auto It = ItBegin; It != ItEnd; ++It) {
+    auto PropRange = It->second->getDeviceRequirements();
+    auto PropIt =
+        std::find_if(PropRange.begin(), PropRange.end(), [&](const auto &Prop) {
+          return Prop->Name == std::string_view("compile_target");
+        });
+    auto AddImg = [&]() { DeviceFilteredImgs.push_back(It->second); };
 
-  std::vector<ur_device_binary_t> UrBinaries(RawImgs.size());
-  for (uint32_t BinaryCount = 0; BinaryCount < RawImgs.size(); BinaryCount++) {
-    UrBinaries[BinaryCount].pDeviceTargetSpec =
-        getUrDeviceTarget(RawImgs[BinaryCount]->DeviceTargetSpec);
+    // Device image has no compile_target property, so it is not even AOT
+    // compiled.
+    if (PropIt == PropRange.end()) {
+      AddImg();
```

Comment:
I'm not sure I understand this. We are adding the image even if it's not AOT compiled? My understanding is that we only want to keep those images that are AOT compiled for the device, so if it's not even AOT compiled, why do we keep it?

Reply:
This is a question of compatibility, i.e. not every AOT-compiled image is compatible with every device, but JIT-compiled images should be compatible with every device.

Of course, there are AOT-only targets like CUDA & HIP, but I assume that filtering for them happens somewhere else, so we don't handle those cases here. I suppose that if comments are rewritten from compatibility point of view, then they should be more clear.

Reply:
Updated the comment in [57d33d0](https://github.com/intel/llvm/pull/14909/commits/57d33d021df23ae79630213a34f1ddd52177e2f2), let me know if that's clearer